File: main2.go
package main

import "fmt"

func punkt(v [][2]string) string {
	n := len(v)

	start := ""
	for i := 0; i < n; i++ {
		ok := true
		for j := 0; j < n; j++ {
			if v[j][1] == v[i][0] {
				ok = false
				break
			}
		}
		if ok {
			start = v[i][0]
			break
		}
	}

	if start == "" {
		return "Ошибка"
	}

	cur := start
	for step := 0; step < n; step++ {
		for i := 0; i < n; i++ {
			if v[i][0] == cur {
				cur = v[i][1]
				break
			}
		}
	}

	if cur == start {
		return "Ошибка"
	}
	return cur
}

func main() {
	var n int
	fmt.Print("Введите количество пар: ")
	fmt.Scan(&n)

	pairs := make([][2]string, n)

	fmt.Println("Введите пары (начало и конец):")
	for i := 0; i < n; i++ {
		fmt.Scan(&pairs[i][0], &pairs[i][1])
	}

	fmt.Println(punkt(pairs))
}

File: main2.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

string punkt(vector<pair<string, string>>& v) {
    int n = v.size();
    
    string start = "";
    for (int i = 0; i < n; ++i) {
        bool ok = true;
        for (int j = 0; j < n; ++j)
            if (v[j].second == v[i].first) ok = false;
        if (ok) { start = v[i].first; break; }
    }
    
    if (start == "") return "Ошибка";
    
    string cur = start;
    for (int step = 0; step < n; step++) {
        for (int i = 0; i < n; i++) {
            if (v[i].first == cur) {
                cur = v[i].second;
                break;
            }
        }
    }
    
    return (cur == start) ? "Ошибка" : cur;
}

int main() {
    setlocale(LC_ALL, "Russian");
    int n;
    cout << "Введите количество пар: ";
    cin >> n;
    
    vector<pair<string, string>> pairs(n);
    
    cout << "Введите пары (начало и конец):" << endl;
    for (int i = 0; i < n; i++) {
        cin >> pairs[i].first >> pairs[i].second;
    }
    cout << punkt(pairs) << endl;
}